packages/gatsby/src/query/graphql-runner.js

```diff
@@ -18,23 +20,96 @@ class GraphQLRunner {
     })
   }
 
-  query(query, context) {
+  async query(query, context, { parentSpan } = {}) {
     const { schema, schemaCustomization } = this.store.getState()
 
-    return graphql(
+    const activity = new SpanTracker(`GraphQL Query`, {
+      parentSpan: parentSpan,
+      tags: {
+        queryName: query.name,
+      },
+    })
+
+    activity.start()
+    const result = await graphql({
       schema,
-      query,
-      context,
-      withResolverContext({
+      source: query.text,
+      rootValue: context,
+      contextValue: withResolverContext({
         schema,
         schemaComposer: schemaCustomization.composer,
-        context,
+        context: {
+          ...context,
+          spanTracker: activity,
+        },
         customContext: schemaCustomization.context,
         nodeModel: this.nodeModel,
       }),
-      context
-    )
+      variableValues: context,
+      fieldResolver: tracingResolver(),
+    })
+    activity.done()
+    return result
+  }
+}
+
+class SpanTracker {
+  constructor(name, { parentSpan, ...args }) {
+    this.parentSpan = tracer.startSpan(name, {
+      childOf: parentSpan,
+      ...args,
+    })
+    this.spans = {}
+  }
+
+  start() {}
+
+  done() {
+    this.parentSpan.finish()
+  }
+
+  createActivity(path, name, args) {
+    let prev = path.prev
+    if (prev && Number.isInteger(prev.key)) {
```

Comment:
https://github.com/graphql/graphql-js/blob/a482d3b24d7a6d4a8027614ab6304e980548e16f/tstypes/jsutils/Path.d.ts

says it can also be a string, i'm unsure where this comes from.